### PR TITLE
hot restart: retry sendmsg() on connection refused

### PR DESCRIPTION
**Pull request by rgs1**
We are tracking sendmsg() failures due when hot restart
happens too quickly (e.g.: a restart loop or the old process
is still pretty new).

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
cc: @jmarantz

**Comment by rgs1** [comment]
cc: @williamsfu99

File: source/server/hot_restarting_base.cc
Comment (jmarantz) [inline source/server/hot_restarting_base.cc:14]: does this need to be 5s? I might have tried 1s to see if that helps.Also, if you are using absl::SleepFor, consider declaring this as absl::Duration.

**Review by jmarantz — COMMENTED** [review]
Looks good generally. Were you able to repro the problem without this, and show that this fixed it?

File: source/server/hot_restarting_base.cc
Comment (rgs1) [reply source/server/hot_restarting_base.cc:14]: I wanted to try something high first to ensure it was good enough, but 1s intervals with a total wait of 10s is probably ok.

I'll move to absl::Duration.

**Comment by rgs1** [comment]
> Looks good generally. Were you able to repro the problem without this, and show that this fixed it?

I'll push it out today/tomorrow and will confirm if it goes away... Just wanted to get the conversation started on how it'll look.

**Comment by jmarantz** [comment]
We're going to wait to see if this can be demonstrated to solve the problem before merging, right?

**Comment by rgs1** [comment]
> We're going to wait to see if this can be demonstrated to solve the problem before merging, right?

Yup, yup -- if it makes through a couple dozen deploys it would be nice to merge. We'll know more next week.

**Comment by jmarantz** [comment]
/wait

ping when we have some results. Also do you want to give a shot at modifying hotrestart_test.sh to remove some of the sleeps to repro without the retries first? It'd be great to test this somehow.

**Comment by rgs1** [comment]
@jmarantz so I think we need another week, but thus far no more crashes during hot restarts (deploys) after this patch. I still need to spend some time getting a repro with the integ tests, but looks promising 🤞 .

**Comment by jmarantz** [comment]
Cool! How often was it crashing before this change?

**Comment by rgs1** [comment]
> Cool! How often was it crashing before this change?

I think we'd pretty much see 1 or 2 occurrences on every deploy... Looking at the data to confirm.

**Comment by rgs1** [comment]
> > Cool! How often was it crashing before this change?
> 
> I think we'd pretty much see 1 or 2 occurrences on every deploy... Looking at the data to confirm.

Ok so just looked at the data, it seems that for the last 2 deploys there were 2 independent cases for which the patch prevented the crash after retrying 8 times.

So maybe we want to go from 10 --> 20 retries? Probably does not warrants making it configurable...

**Comment by rgs1** [comment]
@jmarantz shall we merge this and I can follow-up with the integration tests later?

**Review by jmarantz — APPROVED** [review]
sounds good to me.

@envoyproxy/senior-maintainers

**Comment by mattklein123** [comment]
@rgs1 even if no integration test, can you add a unit test of this functionality? I don't think it would be that hard? Thank you!

/wait

**Comment by rgs1** [comment]
Added a unit test to cover the retry logic, however I need to clean up things a bit to workaround OS X issues, etc.

File: source/server/BUILD
Comment (rgs1) [inline source/server/BUILD:145]: we probably don't need this nor the `#if !defined(__APPLE__)` block below, that was just me developing this in OS X... Once CI is happy can drop them.

File: source/server/hot_restarting_base.cc
Comment (mattklein123) [inline source/server/hot_restarting_base.cc:107]: Why these changes? I think you can use std::array in this case also if you need to change it for some reason?

File: test/server/hot_restarting_base_test.cc
Comment (mattklein123) [inline test/server/hot_restarting_base_test.cc:57]: Hmm. What does this equal in release? Something seems off?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this LGTM modulo small comments!

/wait

File: source/server/hot_restarting_base.cc
Comment (rgs1) [reply source/server/hot_restarting_base.cc:107]: Oh forgot to say, clang on OS X didn't like int foo[]... shall I drop it since we don't care about OS X for hot restart?

File: test/server/hot_restarting_base_test.cc
Comment (rgs1) [reply test/server/hot_restarting_base_test.cc:57]: Yeah fixing this now, figuring out how to handle side-effects when using EXPECT_DEATH

Comment (rgs1) [reply test/server/hot_restarting_base_test.cc:57]: tbh, we don't really need this since the next test properly tests retries. so dropping the retries part, just expecting it to die is good enough.

File: source/server/BUILD
Comment (rgs1) [reply source/server/BUILD]: gone.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by rgs1** [comment]
Error is unrelated:

```
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/dev/.buildkit_qemu_emulator /bin/sh -c echo "d6c40440609a23483f12eb6295b5191e94baf08298a856bab6e15b10c3b82891  /tmp/su-exec.c" > /tmp/checksum     && curl -o /tmp/su-exec.c https://raw.githubusercontent.com/ncopa/su-exec/212b75144bbc06722fbd7661f651390dc47a43d1/su-exec.c     && sha256sum -c /tmp/checksum     && gcc -Wall /tmp/su-exec.c -o/usr/local/bin/su-exec     && chown root:root /usr/local/bin/su-exec     && chmod 0755 /usr/local/bin/su-exec]: exit code: 1

```

**Pull request merged**
